projects/data_cleaning/Project_overview.md

```diff
@@ -0,0 +1,166 @@
+- 5.1 Github Action
+    - 5.1.1 Copy the GitHub action "python-app.yml" from the python template https://github.com/DHI/template-python-library to your own library (make sure it sits in the same folder).
+    - 5.1.2 Change all occurrences of "my_library" in the yml file to your package name "tscleaner"
+    - 5.1.3 Comment out the line with "ruff-action" with "#"
+    - 5.1.4 Commit, push and create a pull request; the tests should now run, verify that they all run before you move on
+- 5.2 Ruff
+    - 5.2.1 Enable the "ruff-action" be removing the "#" you added above
+    - 5.2.2 Commit and push, your actions will probably fail now - inspect the problems by clicking the red cross (did you also get an email?)
+    - 5.2.3 Install "ruff" on your local machine with mamba/conda/pip
+    - 5.2.4 Navigate to your project root folder and run ruff with "ruff ."
+    - 5.2.5 Add `__all__ = ["SpikeCleaner", "FlatPeriodCleaner", "OutOfRangeCleaner", "plot_timeseries"]` to your `__init__.py` file and fix remaining issues until ruff passes
+    - 5.2.6 Commit, push and verify that you action now succeeds
+- 5.3 Black
+    - 5.3.1 Install "black" on your local machine with mamba/conda/pip
+    - 5.3.2 Run black from your project root folder; inspect the differences; commit
+- 5.4 pyproject.toml
+    - Copy the pyproject.toml from the python template https://github.com/DHI/template-python-library (this file will replace your setup.py)
+    - Modify to fit your package
+    - Remove the setup.py
+    - Commit, push and verify that the GitHub action runs
+    - If it fails, you probably forgot some dependencies - go back and fix
+    - [Optional] You should also re-install your local package with ">pip install --upgrade -e ."
+- 5.5 [Optional] Enable black and ruff extensions in VSCode; set black to run on save
+- Create pull request in GitHub and "request review" from your reviewers
+- Get feedback, Adjust code until approval, then merge (and delete branch)
+
+
+## Module 5: Object-oriented design
+
+- Create new branch "oop-dataclasses" (Make sure changes from last module have been merged, and that you start from the main branch)
+- 3.1 Add type hints to all functions and methods. Commit
+- 3.2 Make all the cleaner classes dataclasses.    
+    - remove the init method (not needed anymore)
+    - Check that the notebook still runs and that the classes indeed work as data classes (e.g. have a string representation and support equality testing etc)
+    - Commit
+- 3.3 Make a private module function _print_stats() that prints the number of cleaned values
+    - call from each of the clean methods
+- 3.4 Consider whether it would be better to create a base class BaseCleaner - write down your considerations as a comment in the pull request, refer to specific lines of code
+- (3.5 Implement inheritance and composition where appropriate. TODO ????)
```

Comment:
Create a Pipeline or Workflow with a collection of Cleaners (composition).

Extract common functionality into a BaseCleaner and inherit from this (validating format of input data🤔)